src/KOKKOS/metatensor_system_kokkos.h

```diff
@@ -0,0 +1,157 @@
+/* -*- c++ -*- ----------------------------------------------------------
+   LAMMPS - Large-scale Atomic/Molecular Massively Parallel Simulator
+   https://www.lammps.org/, Sandia National Laboratories
+   LAMMPS Development team: developers@lammps.org
+
+   Copyright (2003) Sandia Corporation.  Under the terms of Contract
+   DE-AC04-94AL85000 with Sandia Corporation, the U.S. Government retains
+   certain rights in this software.  This software is distributed under
+   the GNU General Public License.
+
+   See the README file in the top-level LAMMPS directory.
+------------------------------------------------------------------------- */
+
+#ifndef LMP_METATENSOR_SYSTEM_KOKKOS_H
+#define LMP_METATENSOR_SYSTEM_KOKKOS_H
+
+#include <vector>
+#include <array>
+#include <unordered_set>
+
+#include "pointers.h"
+#include "pair.h"
+#include "neigh_list.h"
+#include "kokkos.h"
+
+#include <metatensor/torch/atomistic.hpp>
+
+
+namespace LAMMPS_NS {
+
+template<class DeviceType>
+struct MetatensorSystemOptionsKokkos {
+    // Mapping from LAMMPS types to metatensor types
+    const int32_t* types_mapping;
+    const Kokkos::View<int32_t*, Kokkos::LayoutRight, DeviceType> types_mapping_kokkos;
+    // interaction range of the model, in LAMMPS units
+    double interaction_range;
+    // should we run extra checks on the neighbor lists?
+    bool check_consistency;
+};
+
+// data for metatensor neighbors lists
+template<class DeviceType>
+struct MetatensorNeighborsDataKokkos {
+    // single neighbors sample containing [i, j, S_a, S_b, S_c]
+    using sample_t = std::array<int32_t, 5>;
+
+    struct SampleHasher {
+        static void hash_combine(std::size_t& seed, const int32_t& v) {
+            seed ^= std::hash<int32_t>()(v) + 0x9e3779b9 + (seed<<6) + (seed>>2);
+        }
+
+        size_t operator()(const sample_t& s) const {
+            size_t hash = 0;
+            hash_combine(hash, s[0]);
+            hash_combine(hash, s[1]);
+            hash_combine(hash, s[2]);
+            hash_combine(hash, s[3]);
+            hash_combine(hash, s[4]);
+            return hash;
+        }
+    };
+
+    // cutoff for this NL in LAMMPS units
+    double cutoff;
+    // options of the NL as requested by the model
+    metatensor_torch::NeighborListOptions options;
+
+    // Below are cached allocations for the LAMMPS -> metatensor NL translation
+    // TODO: report memory usage for these?
+
+    // we keep the set of samples twice: once in `known_samples` to remove
+    // duplicated pairs, and once in `samples` in a format that can be
+    // used to create a torch::Tensor.
+    std::unordered_set<sample_t, SampleHasher> known_samples;
+    std::vector<sample_t> samples;
+    // pairs distances vectors
+    std::vector<std::array<double, 3>> distances_f64;
+    std::vector<std::array<float, 3>> distances_f32;
+};
```

Comment:
This looks identical to the non-kokkos one, why is it duplicated here?

Reply:
True for the second class in this snippet, but not for the first one. I think this one of the very few parts of the code that can be re-used from the original interface, but I'll leave this re-organization up to you